### PR TITLE
feat(linter): implement `no-negated-condition` rule

### PR DESCRIPTION
**Pull request by baseballyama**
implement: https://eslint.org/docs/latest/rules/no-negated-condition

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+impl Rule for NoNegatedCondition {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.kind() {
+            AstKind::IfStatement(if_stmt) => {
+                if !has_else_without_condition(if_stmt) {
+                    return;
+                }
+
+                if is_negated_if(if_stmt) {
+                    ctx.diagnostic(no_negated_condition_diagnostic(node.span()));
```
Comment (Sysix) [inline crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:58]:
```suggestion
                    ctx.diagnostic(no_negated_condition_diagnostic(if_stmt.span));
```

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                }
+            }
+            AstKind::ConditionalExpression(conditional_expr) => {
+                if is_negated_if_conditional(conditional_expr) {
+                    ctx.diagnostic(no_negated_condition_diagnostic(node.span()));
```
Comment (Sysix) [inline crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:63]:
```suggestion
                    ctx.diagnostic(no_negated_condition_diagnostic(conditional_expr.span));
```

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    /// ```js
+    /// if (isReady) {
+    ///   doSomethingElse();
+    /// } else {
+    ///   doSomething();
+    /// }
+    /// ```
+    NoNegatedCondition,
+    style,
+);
```
Comment (Sysix) [inline crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:47]:
```suggestion
    pending,
);
```

We could proably support simple if statements, with `pending` we tell that a auto-fix / suggestion is possible:
https://oxc.rs/docs/guide/usage/linter/rules.html

**Review by Sysix — COMMENTED** [review]
thanks for the work. I commented some possible improvements.  
They are all probably only nitpicking :')

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs (hunk 3 above)
Comment (baseballyama) [reply crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:47]: The original ESLint rule did not implement a fix, so I followed that approach. However, it does seem feasible to implement an automatic fix. I’ll consider to implement this in a another PR.

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs (hunk 1 above)
Comment (baseballyama) [reply crates/oxc_linter/src/rules/eslint/no_negated_condition.rs]: Updated!: https://github.com/oxc-project/oxc/pull/8149/commits/145014607b69f6e49c7c40a7526595d545683276

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs (hunk 2 above)
Comment (baseballyama) [reply crates/oxc_linter/src/rules/eslint/no_negated_condition.rs]: Updated!: https://github.com/oxc-project/oxc/pull/8149/commits/145014607b69f6e49c7c40a7526595d545683276

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs (hunk 3 above)
Comment (baseballyama) [reply crates/oxc_linter/src/rules/eslint/no_negated_condition.rs:48]: Updated!: https://github.com/oxc-project/oxc/pull/8149/commits/145014607b69f6e49c7c40a7526595d545683276

File: crates/oxc_linter/src/rules/eslint/no_negated_condition.rs (hunk 2 above)
Comment (Sysix) [reply crates/oxc_linter/src/rules/eslint/no_negated_condition.rs]: @baseballyama you do not need to call `.span()`. `.span` is enough.
Then you can remove the import `use oxc_span::GetSpan;`. 
Sorry I should have told you what my indentations are.

Comment (baseballyama) [reply crates/oxc_linter/src/rules/eslint/no_negated_condition.rs]: Ohhhh; I didn't see this carefully🙏
Updated: https://github.com/oxc-project/oxc/pull/8149/commits/155d9ea0fc90f90dc2709b36ceaf4b5af8b044a3

And maybe there are same issue on other PRs, so I will check this.

**Review by Sysix — COMMENTED** [review]
👍  you can resolve the discussions (I do not have the rights for that ^^)

**Review by camc314 — CHANGES_REQUESTED** [review]
i think we've already got this rule covered by the unicorn variant?

do you mind porting the test cases to there and seeing if the all pass correctly?

https://github.com/oxc-project/oxc/blob/bd0693bcd01cc826a4717779558e06ce99262732/crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs#L0-L1

**Comment by baseballyama** [comment]
@camc314 

I added tests to the unicorn's rule, and they all passed.
I think it would be better to have a rule on the ESLint side as well, so I rewrote it to forward the processing.
What do you think?

**Comment by Sysix** [comment]
@camc314 personally I like this remapping. Can we do this the same for other rules and maybe fix #8009 ?

**Comment by camc314** [comment]
i haven't looked into https://github.com/oxc-project/oxc/issues/8009.

but we shouldn't have the same code doing the same thing/

That said, it would be OK to map `unicorn/no-negated-condition` to `eslint/no-negated-condition` as we do for some of the TS rules.

To clarify, we would change the following

https://github.com/oxc-project/oxc/blob/3858221f453b8037e5d1d6f5862740db24474345/crates/oxc_linter/src/config/rules.rs#L159-L161

to be something like:

```diff
diff --git a/crates/oxc_linter/src/config/rules.rs b/crates/oxc_linter/src/config/rules.rs
index e661f876f..251026b5f 100644
--- a/crates/oxc_linter/src/config/rules.rs
+++ b/crates/oxc_linter/src/config/rules.rs
@@ -160,6 +160,10 @@ fn transform_rule_and_plugin_name<'a>(
         return (rule_name, "eslint");
     }
 
+    if plugin_name == "unicorn" && rule_name == "no-negated-condition" {
+        return ("no-negated-condition", "eslint");
+    }
+
     (rule_name, plugin_name)
 }
 ```

that way, regardsless of whether you have `eslint/no-negated-condition` OR `no-negated-condition` OR `unicorn/no-negated-condition`, it should turn on the rule correctly.

**Comment by camc314** [comment]
i looked into https://github.com/oxc-project/oxc/issues/8009 briefly and i'm pretty sure it's unrelated to this. it's (possibly - although i couldn't reproduce it) a bug somewhere in our config code

**Comment by baseballyama** [comment]
@camc314 

Thank you for your review! (And other PRs also!)

I updated the PR according to your comment.
https://github.com/oxc-project/oxc/pull/8149/commits/24388b5b54110e06e79ca871c1f96251c4a45786

File: crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs:1]: this file needs to be moved into `rules/eslint`

File: crates/oxc_linter/src/snapshots/unicorn_no_negated_condition.snap
Comment (camc314) [inline crates/oxc_linter/src/snapshots/unicorn_no_negated_condition.snap:1]: this one should be deleted

File: crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs
Comment (baseballyama) [reply crates/oxc_linter/src/rules/unicorn/no_negated_condition.rs:1]: I understand what this process means.
updated!

```rust
    if plugin_name == "unicorn" && rule_name == "no-negated-condition" {
        return ("no-negated-condition", "eslint");
    }
```

File: crates/oxc_linter/src/snapshots/unicorn_no_negated_condition.snap
Comment (baseballyama) [reply crates/oxc_linter/src/snapshots/unicorn_no_negated_condition.snap:1]: Did!

**Review by camc314 — APPROVED** [review]
thanks!

**Pull request merged**